pkg/util/file_indexer.go

```diff
@@ -302,3 +307,311 @@ func WriteFile(newFileMap map[string]FileData, resolvedPath string) error {
+				// 'odo push --context foobar' instead of 'odo push --context ~/foobar' it will NOT work..
+				fileAbsolutePath, err := util.GetAbsPath(fileName)
+				if err != nil {
+					return IndexerRet{}, err
+				}
+				klog.V(4).Infof("Got abs path: %s", fileAbsolutePath)
+				klog.V(4).Infof("Making %s relative to %s", srcPath, fileAbsolutePath)
+
+				// We use "FromSlash" to make this OS-based (Windows uses \, Linux & macOS use /)
+				// we get the relative path by joining the two
+				destFile, err := filepath.Rel(filepath.FromSlash(srcPath), filepath.FromSlash(fileAbsolutePath))
+				if err != nil {
+					return IndexerRet{}, err
+				}
+
+				// Now we get the source file and join it to the base directory.
+				srcFile := filepath.Join(filepath.Base(srcPath), destFile)
+
+				if value, ok := remoteDirectories[destFile]; ok {
+					destFile = value
+				}
+
+				klog.V(4).Infof("makeTar srcFile: %s", srcFile)
+				klog.V(4).Infof("makeTar destFile: %s", destFile)
+
+				// The file could be a regular file or even a folder, so use recursiveTar which handles symlinks, regular files and folders
+				innerRet, err := recursive(directory, filepath.Dir(srcPath), srcFile, filepath.Dir(destPath), destFile, ignoreRules, remoteDirectories, *existingFileIndex, filesystem.DefaultFs{})
+				if err != nil {
+					return IndexerRet{}, err
+				}
+
+				for k, v := range innerRet.NewFileMap {
+					ret.NewFileMap[k] = v
+				}
+
+				ret.FilesChanged = append(ret.FilesChanged, innerRet.FilesChanged...)
+				ret.FilesDeleted = append(ret.FilesDeleted, innerRet.FilesDeleted...)
+				ret.RemoteDeleted = append(ret.RemoteDeleted, innerRet.RemoteDeleted...)
+			}
+		}
+	}
+
+	// find files which are deleted/renamed
+	for fileName, value := range existingFileIndex.Files {
+		if _, ok := ret.NewFileMap[fileName]; !ok {
+			klog.V(4).Infof("Deleting file: %s", fileName)
+
+			if value.RemoteAttribute != "" {
+				ret.FilesDeleted = append(ret.FilesDeleted, value.RemoteAttribute)
+				currentRemote := value.RemoteAttribute
+				ret.RemoteDeleted = append(ret.RemoteDeleted, findRemoteFolderForDeletion(currentRemote, remoteDirectories)...)
+			} else {
+				// check the *absolute* path to the file for glob rules
+				fileAbsolutePath, err := GetAbsPath(filepath.Join(directory, fileName))
+				if err != nil {
+					return ret, errors.Wrapf(err, "unable to retrieve absolute path of file %s", fileName)
+				}
+
+				matched, err := util.IsGlobExpMatch(fileAbsolutePath, ignoreRules)
+				if err != nil {
+					return IndexerRet{}, err
+				}
+				if matched {
+					continue
+				}
+				ret.FilesDeleted = append(ret.FilesDeleted, fileName)
+			}
+		}
+	}
+	return ret, nil
+}
+
+func recursive(directory, srcBase, srcFile, destBase, destFile string, ignoreRules []string, remoteDirectories map[string]string, existingFileIndex FileIndex, fs filesystem.Filesystem) (IndexerRet, error) {
```

Comment:
comments for this function? and also too many arguments passed, can we reduce them?

Reply:
Sure, I will reduce them.